File: sdk/typescript/session.ts
/**
 * 多轮会话 Session
 * ================
 * 用法:
 *   const sess = client.session('你是代码审查助手')
 *   const r1 = await sess.chat('审查这段代码：...')
 *   const r2 = await sess.chat('给出修复方案')
 *   sess.clear()
 */

import OpenAI from 'openai'
import type { LLMClient, ChatOptions } from './client'
import type { ZodType } from './structured'
import { extractRaw } from './structured'

const MAX_HISTORY_TOKENS = 6000

function estimateTokens(text: string): number {
  return Math.max(1, Math.floor(text.length / 4))
}

export class Session {
  private llm: LLMClient
  private system?: string
  private _history: OpenAI.ChatCompletionMessageParam[] = []

  constructor(llm: LLMClient, system?: string) {
    this.llm = llm
    this.system = system
  }

  get turns(): number {
    return Math.floor(this._history.length / 2)
  }

  get history(): OpenAI.ChatCompletionMessageParam[] {
    return [...this._history]
  }

  clear(): void {
    this._history = []
  }

  private trimHistory(): void {
    while (this._history.length >= 2) {
      const total = this._history.reduce((sum, m) => {
        const content = typeof m.content === 'string' ? m.content : JSON.stringify(m.content)
        return sum + estimateTokens(content)
      }, 0)
      if (total <= MAX_HISTORY_TOKENS) break
      this._history = this._history.slice(2)
    }
  }

  private append(role: 'user' | 'assistant', content: string): void {
    this._history.push({ role, content })
    this.trimHistory()
  }

  async chat(prompt: string, opts: Omit<ChatOptions, 'system' | 'history'> = {}): Promise<string> {
    const response = await this.llm.chat(prompt, {
      ...opts,
      system: this.system,
      history: this._history,
    })
    // Handle both dict (with cache) and string (backward compatible)
    const reply = typeof response === 'string' ? response : response.content
    this.append('user', prompt)
    this.append('assistant', reply)
    return reply
  }

  async *chatStream(
    prompt: string,
    opts: Omit<ChatOptions, 'system' | 'history'> = {},
  ): AsyncGenerator<string> {
    const chunks: string[] = []
    for await (const chunk of this.llm.chatStream(prompt, {
      ...opts,
      system: this.system,
      history: this._history,
    })) {
      chunks.push(chunk)
      yield chunk
    }
    const fullReply = chunks.join('')
    this.append('user', prompt)
    this.append('assistant', fullReply)
  }

  async chatStructured<T>(
    prompt: string,
    schema: ZodType<T>,
    opts: Omit<ChatOptions, 'system' | 'history'> = {},
  ): Promise<T> {
    const [result, rawText] = await extractRaw(this.llm, prompt, schema, {
      ...opts,
      system: this.system,
      history: this._history,
    })
    this.append('user', prompt)
    this.append('assistant', rawText)
    return result
  }
}

File: sdk/python/llm_sdk/session.py
"""
多轮会话 Session
===============
用法:
    session = client.session(system="你是代码审查助手")
    r1 = session.chat("审查这段代码：...")
    r2 = session.chat("给第一个问题一个修复方案")
    session.clear()
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Iterator, Type, TypeVar

if TYPE_CHECKING:
    from .client import LLMClient

T = TypeVar("T")

# history 超过此 token 估算值时，截断最早的非 system 轮
_MAX_HISTORY_TOKENS = 6000


def _estimate_tokens(text: str) -> int:
    """粗估 token 数（1 token ≈ 4 字符）"""
    return max(1, len(text) // 4)


class Session:
    """
    封装多轮对话 history，业务代码只需调用 chat()，
    无需手动维护 messages 列表。
    """

    def __init__(self, llm: "LLMClient", system: str | None = None):
        self._llm = llm
        self._system = system
        self._history: list[dict] = []  # 不含 system，只含 user/assistant

    # ── 状态查询 ─────────────────────────────────────────────────────────────

    @property
    def turns(self) -> int:
        """已完成的对话轮数（1 user + 1 assistant = 1 轮）"""
        return len(self._history) // 2

    @property
    def history(self) -> list[dict]:
        """只读 history 快照"""
        return list(self._history)

    def clear(self) -> None:
        """清空 history，保留 system prompt"""
        self._history = []

    # ── 内部工具 ─────────────────────────────────────────────────────────────

    def _trim_history(self) -> None:
        """当 history 过长时，移除最早的完整轮（user+assistant 对）"""
        while len(self._history) >= 2:
            total = sum(
                _estimate_tokens(
                    m["content"] if isinstance(m["content"], str) else str(m["content"])
                )
                for m in self._history
            )
            if total <= _MAX_HISTORY_TOKENS:
                break
            # 移除最早一轮（user + assistant）
            self._history = self._history[2:]

    def _append(self, role: str, content: str) -> None:
        self._history.append({"role": role, "content": content})
        self._trim_history()

    # ── 公开接口 ─────────────────────────────────────────────────────────────

    def chat(
        self,
        prompt: str,
        *,
        image_url: str | None = None,
        image_path: str | None = None,
        capability: str | None = None,
        model: str | None = None,
        cache: bool = True,
        **kwargs,
    ) -> str:
        """发送一轮对话，自动携带历史"""
        response = self._llm.chat(
            prompt,
            system=self._system,
            history=self._history,
            image_url=image_url,
            image_path=image_path,
            capability=capability,
            model=model,
            cache=cache,
            **kwargs,
        )
        # Handle both dict (with cache) and string (backward compatible)
        if isinstance(response, dict):
            reply = response.get("content", "")
        else:
            reply = response
        self._append("user", prompt)
        self._append("assistant", reply)
        return reply

    def chat_stream(
        self,
        prompt: str,
        *,
        capability: str | None = None,
        model: str | None = None,
        **kwargs,
    ) -> Iterator[str]:
        """流式对话，流结束后自动追加到 history"""
        chunks: list[str] = []
        for chunk in self._llm.chat_stream(
            prompt,
            system=self._system,
            history=self._history,
            capability=capability,
            model=model,
            **kwargs,
        ):
            chunks.append(chunk)
            yield chunk
        full_reply = "".join(chunks)
        self._append("user", prompt)
        self._append("assistant", full_reply)

    def chat_structured(
        self,
        prompt: str,
        schema: Type[T],
        *,
        capability: str | None = None,
        model: str | None = None,
        **kwargs,
    ) -> T:
        """结构化输出 + 自动维护 history"""
        from .structured import chat_structured as _cs

        result, reply_text = _cs(
            self._llm,
            prompt,
            schema,
            system=self._system,
            history=self._history,
            capability=capability,
            model=model,
            _return_raw=True,
            **kwargs,
        )
        self._append("user", prompt)
        self._append("assistant", reply_text)
        return result
